README.md

```diff
@@ -1,7 +1,154 @@
+This library is intended to be used for verifying string-manipulating Rust programs via Lean.
+
+The Lean implementation is based on the description of the Rust functions, which are published at https://doc.rust-lang.org/std.
+The library includes:
+- Definitions of functions equivalent to those from the Rust standard library
+- Proofs that these definitions are correct (i.e., match our understanding of the Rust original)
+- Supporting lemmas and theorems for proof construction
+
+## Installation
+
+To use the library, add the following lines to your `lakefile.lean`:
+
+` require rust-lean-models from git "https://github.com/model-checking/rust-lean-models" `
+
+Then do `lake update` in the terminal.
+
+## Usage
+### Translating Rust program 
+
+For any Rust built-in functions which are used in user code , map it with 
+the corresponding function name in the library (see the Tables below).
+
+### Proof Tutorial
+We demonstrate some proof techniques in Proof_Tutorial.lean through two simple programs that compute
+the longest common prefix and longest common substring of the two input strings. 
+More examples can be found in Proof_Example.lean
+
+## Implementation
+
+
+### Recursive function definition
+For each Rust function, we provide a recursive Lean function. Implementing 
+the equivalent functions in Lean recursively enables user to construct 
+induction proofs conveniently. The Lean function has the same as the Rust original, 
+except when the Rust name  clashes with a Lean keyword. In case of a clash, a Rust function 'func_name' 
+is renamed to `rust_func_name` in Lean.
+
+Implementing a function recursively may requires some extra parameters.
+In those cases, first, we implement an auxiliary function (name: `func_name_aux`) which is defined 
+recursively with the parameters, then we define the function `func_name` based on the auxiliary function 
+with initial values for the parameters. 
+For example, `char_indices` is defined based on `char_indices_aux` as 
+`def char_indices (s: Str):= char_indices_aux s 0`.
+
+For Rust functions involving the `Pattern` type, we implement two recursive sub-functions 
+(name: `func_name_char_filter` and  `func_name_substring`) which replace the `Pattern` type 
+in the input by either a char filter (or type`Char → Bool`) or a `List Char`. Then we define 
+the function `func_name` based on these two sub-functions by matching the input of `Pattern` type.
+For example, `split` is defined by two sub-functions `split_char_filter` and `split_substring` as: 
+
+```
+def split (s: Str) (p: Pattern) := match p with
+| Pattern.SingleChar c => split_char_filter s (fun x => x = c)
+| Pattern.ListChar l => split_char_filter s (fun x => x ∈ l)
+| Pattern.FilterFunction f => split_char_filter s f
+| Pattern.WholeString s1 => split_substring s s1
+```
+All recursive implementations are proven to be "correct" in the sense that they are consistent with
+the Rust original (see below for more details).
```

Comment:
I would reword this to "are consistent with the descriptions of the Rust versions".